File: social-flow-backend/shared/libraries/go/auth/oauth.go
// ========================================
// File: oauth.go
// ========================================
package auth

import (
	"context"
	"errors"
	"golang.org/x/oauth2"
	"golang.org/x/oauth2/github"
	"golang.org/x/oauth2/google"
)

// OAuthConfig stores credentials for providers
type OAuthConfig struct {
	GoogleClientID     string
	GoogleClientSecret string
	GoogleRedirectURL  string

	GitHubClientID     string
	GitHubClientSecret string
	GitHubRedirectURL  string
}

type OAuthManager struct {
	google *oauth2.Config
	github *oauth2.Config
}

func NewOAuthManager(cfg *OAuthConfig) *OAuthManager {
	return &OAuthManager{
		google: &oauth2.Config{
			ClientID:     cfg.GoogleClientID,
			ClientSecret: cfg.GoogleClientSecret,
			RedirectURL:  cfg.GoogleRedirectURL,
			Scopes:       []string{"email", "profile"},
			Endpoint:     google.Endpoint,
		},
		github: &oauth2.Config{
			ClientID:     cfg.GitHubClientID,
			ClientSecret: cfg.GitHubClientSecret,
			RedirectURL:  cfg.GitHubRedirectURL,
			Scopes:       []string{"user:email"},
			Endpoint:     github.Endpoint,
		},
	}
}

func (o *OAuthManager) GetLoginURL(provider, state string) (string, error) {
	switch provider {
	case "google":
		return o.google.AuthCodeURL(state), nil
	case "github":
		return o.github.AuthCodeURL(state), nil
	default:
		return "", errors.New("unsupported provider")
	}
}

func (o *OAuthManager) ExchangeCode(ctx context.Context, provider, code string) (*UserClaims, error) {
	var conf *oauth2.Config
	switch provider {
	case "google":
		conf = o.google
	case "github":
		conf = o.github
	default:
		return nil, errors.New("unsupported provider")
	}

	token, err := conf.Exchange(ctx, code)
	if err != nil {
		return nil, err
	}

	// Normally, you would call provider’s userinfo endpoint.
	// For brevity, we’ll fake user extraction.
	return &UserClaims{
		UserID:   "oauth-user",
		Username: provider + "_user",
		Roles:    []string{"user"},
	}, nil
}

File: social-flow-backend/shared/libraries/node/auth/oauth.js
/**
 * OAuth2 integration for Google & GitHub
 */

const { google } = require('googleapis');
const axios = require('axios');
const config = require('./config');

module.exports = {
  // Google OAuth2
  getGoogleAuthUrl: () => {
    const oauth2Client = new google.auth.OAuth2(
      config.oauth.googleClientId,
      config.oauth.googleClientSecret,
      config.oauth.redirectUri
    );

    return oauth2Client.generateAuthUrl({
      access_type: 'offline',
      scope: ['profile', 'email'],
    });
  },

  getGoogleUser: async (code) => {
    const oauth2Client = new google.auth.OAuth2(
      config.oauth.googleClientId,
      config.oauth.googleClientSecret,
      config.oauth.redirectUri
    );

    const { tokens } = await oauth2Client.getToken(code);
    oauth2Client.setCredentials(tokens);

    const oauth2 = google.oauth2({
      auth: oauth2Client,
      version: 'v2',
    });

    const { data } = await oauth2.userinfo.get();
    return data;
  },

  // GitHub OAuth2
  getGithubAuthUrl: () =>
    `https://github.com/login/oauth/authorize?client_id=${config.oauth.githubClientId}&redirect_uri=${config.oauth.redirectUri}&scope=user:email`,

  getGithubUser: async (code) => {
    const tokenRes = await axios.post(
      'https://github.com/login/oauth/access_token',
      {
        client_id: config.oauth.githubClientId,
        client_secret: config.oauth.githubClientSecret,
        code,
        redirect_uri: config.oauth.redirectUri,
      },
      { headers: { Accept: 'application/json' } }
    );

    const accessToken = tokenRes.data.access_token;
    const userRes = await axios.get('https://api.github.com/user', {
      headers: { Authorization: `Bearer ${accessToken}` },
    });

    return userRes.data;
  },
};
